Feature/initial commits

SolomonShorser-OICR opened the pull request:
Initial commit of work for COSMIC Update.
NOTE: this PR depends on https://github.com/reactome/release-common-lib/pull/4 so this should not be merged until the release-common-lib PR is merged.

SolomonShorser-OICR committed:
Initial commit

SolomonShorser-OICR committed:
POM and gitignore

SolomonShorser-OICR committed:
Main java files and log4j file

SolomonShorser-OICR committed:
Main java files and log4j file

SolomonShorser-OICR committed:
Ignore target output directory

SolomonShorser-OICR committed:
Ignore target output directory

SolomonShorser-OICR committed:
some cleanup

SolomonShorser-OICR committed:
some cleanup

SolomonShorser-OICR committed:
Some cleanup.

SolomonShorser-OICR committed:
Some cleanup.

SolomonShorser-OICR committed:
Some cleanup

SolomonShorser-OICR committed:
Some cleanup

SolomonShorser-OICR committed:
Rebasing

rebase in progress; onto ea811b2
 You are currently rebasing branch 'feature/initial-commits' on 'ea811b2'.

SolomonShorser-OICR committed:
Merge branch 'feature/initial-commits' of github.com:reactome/release-cosmic-update into feature/initial-commits

SolomonShorser-OICR committed:
Cleanup & refactoring

SolomonShorser-OICR committed:
Refactoring & Documentation.

SolomonShorser-OICR committed:
Consider testMode...

SolomonShorser-OICR committed:
Config file ; Cleanup & Refactoring.

SolomonShorser-OICR committed:
Some cleanup.

SolomonShorser-OICR committed:
Add command-line functionality

SolomonShorser-OICR committed:
Fixed spelling error.

SolomonShorser-OICR committed:
Use COSMICFileRetriever

SolomonShorser-OICR committed:
Cleanup

SolomonShorser-OICR committed:
call downloadFiles()

SolomonShorser-OICR committed:
Add URLs for COSMIC files.

SolomonShorser-OICR committed:
Unzip downloaded files.

SolomonShorser-OICR committed:
used correct fileAge; added some javadoc.

SolomonShorser-OICR committed:
Migrate GUnzipCallable from COSMIC Update to release-common-lib

SolomonShorser-OICR committed:
Ignore download files

SolomonShorser-OICR committed:
Consistent file names.

SolomonShorser-OICR committed:
Changed monitor object.

SolomonShorser-OICR committed:
get personID from config.

SolomonShorser-OICR committed:
Ignore csv, reports, logs

SolomonShorser-OICR committed:
Minor refactoring

SolomonShorser-OICR committed:
Private constructor ; better error message.

SolomonShorser-OICR committed:
Log to file

SolomonShorser-OICR committed:
output to reports directory

SolomonShorser-OICR committed:
Move method to util class

SolomonShorser-OICR committed:
Basic unit tests

SolomonShorser-OICR committed:
Minor refactoring & cleanup.

SolomonShorser-OICR committed:
Unit tests

SolomonShorser-OICR committed:
Unit tests.

SolomonShorser-OICR committed:
Unit testing

SolomonShorser-OICR committed:
Allow custom report dir; add datetime to report filename.

SolomonShorser-OICR committed:
assert report files exist.

SolomonShorser-OICR committed:
Include convenient search URL in report.

SolomonShorser-OICR commented:
@cookersjs @jweiser No reason you can't start a review on this, even though it's waiting on another PR to be ready for merging.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,177 @@
+	{
+		this.mutationIDs = mutationIDs;
+	}
+	public String getCosvIdentifier()
```
What is Cosv? Cosmic Variant?

SolomonShorser-OICR replied:
COSV is "Genomic mutation identifier", according to COSMIC.

See: https://cancer.sanger.ac.uk/cosmic/variant-updates

jweiser replied:
Maybe `getCOSVIdentifier()` since COSMIC uses all capitals?

jweiser replied:
^-- do you think this is worthwhile?

SolomonShorser-OICR committed:
cleanup & refactoring.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+			
+			updateIdentifierObject(adaptor, COSMICIdentifierUpdater.updateWithNewCOSV, identifierObject, this.getCosvIdentifier());
+		}
+		// If no COSV identifier was found, update using the suggested prefix (determined computationally).
```
Could this comment have examples of the legacy prefixes? If its a long list of examples, then at least a couple would be good

SolomonShorser-OICR replied:
"COSM" is the legacy prefix.

SolomonShorser-OICR replied:
https://github.com/reactome/release-cosmic-update/blob/feature/initial-commits/src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java#L34

SolomonShorser-OICR replied:
But also, some COSMIC identifiers in Reactome have _no_ prefix, they are numeric. Which is what this code is trying to fix.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+			String currentIdentifier = (String) identifierObject.getAttributeValue(ReactomeJavaConstants.identifier);
+			// If the current identifier already begins with "C" then leave it alone.
+			// This code is for updating numeric identifiers that have a suggested prefix.
+			if (!currentIdentifier.toUpperCase().startsWith("C"))
```
Could this check be for if the currentIdentifier starts with an integer? That way it stays in sync with the commenting, and in case in the future if they introduce non-C-starting identifiers (introducing KOSMIC!), it can handle that

SolomonShorser-OICR replied:
Checking that the first character is numeric would require a regex (or parsing a string to numeric) vs. checking that the first character is "C". Checking the first character of a string is much simpler than evaluating a regular expression or parsing a string to a number (which could also involve regular expressions).

jweiser replied:
I understand the point about regexes, but I think checking for a number would be better since the code will attempt to prepend "COSM" anything that doesn't start with a C.  This isn't a problem now, but it is vulnerable to wrongly pretending "COSM" if a different identifier exists later.

Checking the precise case of if it is an integer seems safer despite a slightly more expensive check so we know it only works with that kind of identifier.

SolomonShorser-OICR replied:
COSMIC seems pretty committed to their "COS-" notation. I think we can change the code to handle other identifiers once COSMIC starts using a different notation system.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+	 * <code>identifierObject</code> (which must be an InstanceEdit) will also have <code>modifiedForCOSMICUpdate</code> added to its <code>modified</code> list.
+	 * The display name of <code>identifierObject</code> will also be regenerated to reflect changes in <code>identifierValue</code>.
+	 * @param adaptor
+	 * @param modifiedForCOSMICUpdate An InstanceEdit the explains why an instance was modified.
```
typo, 'that' explains why

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+	// Private constructor to prevent instantiation of utility class
+	private COSMICUpdateUtil()
+	{
+		// ...no-op
```
What does this mean?

SolomonShorser-OICR replied:
"no-op" means "no operation". I put the comment there to make it 100% clear that it was intentional that there is no code in this method.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+	 * Validate the identifiers in the database by comparing them to the identifiers in the file.
+	 * @param updaters A map of updaters, keyed by COSMIC identifier.
+	 * @param COSMICFusionExportFile The path to the COSMIC Fusion Export file.
+	 * @param COSMICMutationTrackingFile The path o the COSMIC Mutation Tracking file.
```
typO

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+		Map<String, COSMICIdentifierUpdater> updates = new HashMap<>();
+		
+		// Create the reports directory if it's missing.
+		if (!Files.exists(Paths.get(COSMICUpdateUtil.reportsDirectoryPath)))
+		{
+			Files.createDirectories(Paths.get(COSMICUpdateUtil.reportsDirectoryPath));
+		}
```
Any concerns if the directory does exist?

SolomonShorser-OICR replied:
If it already exists, then nothing needs to be done. The reports will be written to it.

cookersjs replied:
I guess I meant is there any situation where the directory existing could cause issues, such as if there was old data there

SolomonShorser-OICR replied:
The report files have a date and time in the filename so there should be no problems with older files existing in the report directory.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+				}
+				
+				// If the identifier starts with C it's not a numeric identifier.
+				if (identifier.startsWith("C"))
```
Same as https://github.com/reactome/release-cosmic-update/pull/1/files#r552224756 - the check itself is for the identifier starting with 'C' (which is obviously valid), but the comment is regarding it not being numeric

SolomonShorser-OICR replied:
Checking that the first character is numeric would require a regex (or parsing a string to numeric) vs. checking that the first character is "C". Checking the first character of a string is much simpler than evaluating a regular expression or parsing a string to a number (which could also involve regular expressions).

jweiser replied:
Similar to https://github.com/reactome/release-cosmic-update/pull/1/files#r568852047

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+				// If the identifier starts with C it's not a numeric identifier.
+				if (identifier.startsWith("C"))
+				{
+					// Consider: is it possible that we might get 1:n mapping for some identifier?
```
What would happen in the code if that were the case? Would it throw an error, or would it just overwrite the existing one?

SolomonShorser-OICR replied:
I don't think I've found an instance of this yet...

jweiser replied:
Maybe a check and a logging statement for this case would be good so we're aware if there are many for one?

SolomonShorser-OICR replied:
Code now handles the possibility that > 1 objects might use the same COSMIC identifier.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+			}
+			else
+			{
+				nonEWASPrinter.printRecord(identifier, ewas.toString());
```
Won't printing a GKInstance to string also include the square brackets? Nothing wrong with that, but I'm not sure how the file produced is used and if it is anything consumed by users I think we tend to avoid the full instance name

SolomonShorser-OICR replied:
Here's an example of what it looks like:
```
[CandidateSet:1602342] FBXW7 R505/387C mutant [nucleoplasm]
```
The file is a CSV file. Lisa has already seen this in the report and has not had any problems with it. The purpose of this report is to let the curators know that there are non-EWAS objects identified with COSMIC identifiers.

cookersjs replied:
Ah OK, I figured it was internal use but I wasn't sure. Sounds good!

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+		while (!foundMismatchedRefSequence && i < modResidues.size())
+		{
+			GKInstance modResidue = modResidues.get(i);
+			if (modResidue.getSchemClass().getName().contains(ReactomeJavaConstants.FragmentReplacedModification)
```
I think there is a `isa()` method for SchemaClass objects that would allow you to remove the `getName().contains()` part of this statement

SolomonShorser-OICR replied:
I may look into this...

jweiser replied:
You could use an iterator here, too, if you want so `int i` isn't needed.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+				|| modResidue.getSchemClass().getName().contains(ReactomeJavaConstants.FragmentInsertionModification))
+			{
+				GKInstance residueRefSequence = (GKInstance) modResidue.getAttributeValue(ReactomeJavaConstants.referenceSequence);
+				foundMismatchedRefSequence = !residueRefSequence.getDBID().equals(refSequence.getDBID());
```
I think I find this line a little tricky to read because its a boolean value that is based on a ! - looking at the code, it doesn't really make sense to re-organize everything, but perhaps a helper method for legibility?

SolomonShorser-OICR replied:
I'm not sure what the problem is.

cookersjs replied:
I was just saying that its a boolean value created from a !, so I don't know, I guess like saying `boolean colourIsBlue = !object.getColour().isRed()`-- nothing wrong with it, just that reading it took a moment for me to unpack how the statement worked

SolomonShorser-OICR replied:
Ok. I don't think there's any further way to simplify this, so I'm going to leave it as it is.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+		}
+		
+		@SuppressWarnings("unchecked")
+		Collection<GKInstance> cosmicObjects = adaptor.fetchInstanceByAttribute(ReactomeJavaConstants.DatabaseIdentifier, ReactomeJavaConstants.referenceDatabase, " = ", cosmicRefDB.getAttributeValue(ReactomeJavaConstants.DB_ID));
```
I didn't realize you could do this - so you are searching for instances that have the COSMIC refDB, and rather than feeding it the refDB instance, you can just feed it the DBID? Thats neat!

SolomonShorser-OICR replied:
`fetchInstanceByAttribute` is a very useful function.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+				// Include a COSMIC Search URL for the identifier in the report, to make it easier for Curators to follow up on identifiers that might need attention.
+				String url;
+				String identifierForUrl = "";
+				if (record.getIdentifier().startsWith("C"))
```
Same as https://github.com/reactome/release-cosmic-update/pull/1/files#r552224756 - though there isn't any commenting about filtering by numeric identifiers so its less of an issue

SolomonShorser-OICR replied:
Checking that the first character is numeric would require a regex (or parsing a string to numeric) vs. checking that the first character is "C". Checking the first character of a string is much simpler than evaluating a regular expression or parsing a string to a number (which could also involve regular expressions).

jweiser replied:
Similar to here https://github.com/reactome/release-cosmic-update/pull/1/files#r568852047.

Since this kind of check occurs more than 2 times, could a static helper function be added for it?

SolomonShorser-OICR replied:
I think this expression is simple enough that a separate function is not needed.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+						identifierForUrl = record.getIdentifier();
+					}
+				}
+				url = "https://cancer.sanger.ac.uk/cosmic/search?q=" + identifierForUrl;
```
Could this URL be a global variable? I know it only appears once, but I think cosmic as a resource is growing quite a bit and I wouldn't be surprised if URLs like this change in the future as they continue to update the website, so it would be good for finding it more quickly

jweiser replied:
I'm kinda in favour of this as well.  Maybe a `public static final` constant in the utils class?

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,227 @@
+				Main.COSMICMutationTracking = props.getProperty("pathToMutationTrackingFile", "./CosmicMutationTracking.tsv");
+				Main.COSMICMutantExportURL = props.getProperty("urlToMutantExportFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/CosmicMutantExport.tsv.gz");
+				Main.COSMICFusionExportURL = props.getProperty("urlToFusionExportFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/CosmicFusionExport.tsv.gz");
+				Main.COSMICMutationTrackingURL = props.getProperty("urlToMutationTrackingFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92//CosmicMutationTracking.tsv.gz");
```
There is a double slash after `v92` in this URL

SolomonShorser-OICR replied:
So there is!

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,227 @@
+				Main.COSMICMutantExport = props.getProperty("pathToMutantExportFile", "./CosmicMutantExport.tsv");
+				Main.COSMICFusionExport = props.getProperty("pathToFusionExportFile", "./CosmicFusionExport.tsv");
+				Main.COSMICMutationTracking = props.getProperty("pathToMutationTrackingFile", "./CosmicMutationTracking.tsv");
+				Main.COSMICMutantExportURL = props.getProperty("urlToMutantExportFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/CosmicMutantExport.tsv.gz");
+				Main.COSMICFusionExportURL = props.getProperty("urlToFusionExportFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/CosmicFusionExport.tsv.gz");
+				Main.COSMICMutationTrackingURL = props.getProperty("urlToMutationTrackingFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92//CosmicMutationTracking.tsv.gz");
```
Could the current COSMIC version (v92) be a default value while also allowing the version to be configurable as a property?

SolomonShorser-OICR replied:
Possibly.

SolomonShorser-OICR replied:
On second thought, these URLs are default values - the values in the Java code only get used if there is no value in the properties file. The URLs _really_ should be coming from the properties file. Maybe I'll remove the URLs from here completely.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,227 @@
+				Main.COSMICMutantExport = props.getProperty("pathToMutantExportFile", "./CosmicMutantExport.tsv");
+				Main.COSMICFusionExport = props.getProperty("pathToFusionExportFile", "./CosmicFusionExport.tsv");
+				Main.COSMICMutationTracking = props.getProperty("pathToMutationTrackingFile", "./CosmicMutationTracking.tsv");
+				Main.COSMICMutantExportURL = props.getProperty("urlToMutantExportFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/CosmicMutantExport.tsv.gz");
+				Main.COSMICFusionExportURL = props.getProperty("urlToFusionExportFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/CosmicFusionExport.tsv.gz");
+				Main.COSMICMutationTrackingURL = props.getProperty("urlToMutationTrackingFile", "https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92//CosmicMutationTracking.tsv.gz");
```
Since this URL appears multiple times, could it be a global variable? Also related to https://github.com/reactome/release-cosmic-update/pull/1/files#r552245526

SolomonShorser-OICR replied:
Which URL appears multiple times? To my knowledge, `https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/CosmicMutationTracking.tsv.gz` only appears once. `https://cancer.sanger.ac.uk/cosmic/search?q=` is a different URL.

cookersjs replied:
I should have said partial URL: `https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/`

SolomonShorser-OICR replied:
Ah ok. Yes, that part _could_ be extracted. Not sure I will...

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,227 @@
+		}
+		catch (Exception e)
+		{
+			logger.error("Error occurred while trying to download a file! " + e.getMessage(), e);
```
I assume this is the case but will the error message include the name of the file that caused the error?

SolomonShorser-OICR replied:
The name of the file being downloaded is logged before the download begins. So if the logs says "downloading file: SomeFile.tgz" and then says "downloading file: SomeOtherFile.tgz" and then says "Error occurred while trying to download a file!", it's a good guess that SomeFile.tgz downloaded OK and SomeOtherFile.tgz didn't download OK.

cookersjs commented on src/main/resources/config.properties:
```diff
@@ -0,0 +1,9 @@
+db.host=localhost
+db.user=root
+db.password=root
+db.name=gk_central
```
An absolute nitpick - but do we consider gk_central public information? Maybe `curationDBName` or something? I don't know if it matters, but wanted to point it out

cookersjs reviewed (commented):
I didn't find any blockers, so if this wasn't a _draft_ PR, I'd be approving. But it is a _draft_ PR, so I'm not!

SolomonShorser-OICR committed:
minor cleanups

 - move config values to properties file and remove defaults to avoid
confusion
 - validate inputs

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+					+ this.getSuggestedPrefix() + "; "
+					+ this.isValid() + "; "
+					+ this.getCosvIdentifier() + "; ("
+					+ this.getMutationIDs().stream().reduce("", (s, t) -> s + t + ",") + ")" + "]";
```
Is the reduce creating a comma delimited String?  If so, it might be clearer to use `.collect(Collectors.joining(","))` instead of `reduce()` or String.join(",", this.getMutationIDs()).

jweiser replied:
This might be easier using `String.join(",", this.getMutationIDs())` in place of `.stream().reduce()`.

SolomonShorser-OICR committed:
Added a Jenkins file.

cookersjs commented on cosmic-update.jenkinsfile:
```diff
@@ -0,0 +1,147 @@
+import groovy.json.JsonSlurper
+def currentRelease
+def previousRelease
```
Actually, we've switched the Release pipeline JFs to use the shared groovy library. This allows the current/previous release versions to be retrieved within a stage so 'executing from stage' functionality actually works.

[Update-stable-ids Jenkinsfile](https://github.com/reactome/release-update-stable-ids/blob/develop/Jenkinsfile)

SolomonShorser-OICR replied:
This file predates the shared library. I checked it in yesterday, but I think I authored it... a while ago! I'll make some changes to be more consistent with the other Jenkinsfiles.

cookersjs commented on cosmic-update.jenkinsfile:
```diff
@@ -0,0 +1,147 @@
+			{
+				script
+				{
+					// Get current release number from directory
+					currentRelease = (pwd() =~ /Releases\/(\d+)\//)[0][1];
+					previousRelease = (pwd() =~ /Releases\/(\d+)\//)[0][1].toInteger() - 1;
+					// This queries the Jenkins API to confirm that the most recent build of 'UniProt Update' was successful.
+					def uniprotStatusUrl = httpRequest authentication: 'jenkinsKey', validResponseCodes: "${env.VALID_RESPONSE_CODES}", url: "${env.JENKINS_JOB_URL}/job/${currentRelease}/job/Pre-Slice/job/UniProtUpdate/lastBuild/api/json"
+					if (uniprotStatusUrl.getStatus() == 404)
+					{
+						error("UniProt Update has not yet been run. Please complete a successful build.")
+					}
+					else
+					{
+						def uniprotStatusJson = new JsonSlurper().parseText(uniprotStatusUrl.getContent())
+						if (uniprotStatusJson['result'] != "SUCCESS")
+						{
+							error("Most recent UniProt Update build status: " + uniprotStatusJson['result'] + ". Please complete a successful build.")
+						}
+					}
```
This whole block can be replaced by the shared library's `checkUpstreamBuildSucceeded("UniProtUpdate")` method

SolomonShorser-OICR replied:
I've update the Jenkinsfile to make use of the shared library.

cookersjs reviewed (commented):
Looks like the JF was done using the old methods, I recommend checking out one of the recently updated JFs on any of the release steps and incorporating those methods. Should increase legibility and also reduce the amount of code in each JF

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+	private static GKInstance updateWithNewCOSV;
+	private static GKInstance updatePrependCOSM;
+	private String identifier;
+	private Long dbID;
```
No need to change it but something that has bitten me in the past is comparing `Long` objects to see if they are equal with `==` and getting `false` when the values were the same because it checked reference equality (with `Long` objects you need to use `.equals()` for value equality).  

I always use primitive `long` where I can so it's not an issue.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+public class COSMICIdentifierUpdater implements Comparable<COSMICIdentifierUpdater>
+{
+	private static final Logger logger = LogManager.getLogger();
+	private static GKInstance updateWithNewCOSV;
```
I wasn't sure what type of instance this was until further down.  Could you rename this to `instanceEditForNewCOSVUpdate`?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+{
+	private static final Logger logger = LogManager.getLogger();
+	private static GKInstance updateWithNewCOSV;
+	private static GKInstance updatePrependCOSM;
```
Could this be renamed `instanceEditForPrependCOSMUpdate` so it's explicit it is an `InstanceEdit`?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+	static final String COSMIC_FUSION_PREFIX = "COSF";
+	private static final Logger logger = LogManager.getLogger();
+	private static String dateSuffix;
+	private static String reportsDirectoryPath = "reports";
```
Maybe have this as type `Path` here so there is no need to convert it later on?

SolomonShorser-OICR replied:
It is used as a prefix to to a path to a file, so a String is more appropriate.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+		Map<String, COSMICIdentifierUpdater> updates = new HashMap<>();
+		
+		// Create the reports directory if it's missing.
+		if (!Files.exists(Paths.get(COSMICUpdateUtil.reportsDirectoryPath)))
```
This can be removed if you want,

```
public static Path createDirectories(Path dir,
                     FileAttribute<?>... attrs)
                              throws IOException

Creates a directory by creating all nonexistent parent directories first. Unlike the createDirectory method, an exception is not thrown if the directory could not be created because it already exists. 
```

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+			Files.createDirectories(Paths.get(COSMICUpdateUtil.reportsDirectoryPath));
+		}
+		
+		try(CSVPrinter nonEWASPrinter = new CSVPrinter(new FileWriter(COSMICUpdateUtil.reportsDirectoryPath + "/nonEWASObjectsWithCOSMICIdentifiers_"+dateSuffix+".csv"), CSVFormat.DEFAULT.withHeader("COSMIC identifier", "Referred-to-by CadidateSet (or other non-EWAS)"));
```
Typo 'Candidate'

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+		try(CSVPrinter nonEWASPrinter = new CSVPrinter(new FileWriter(COSMICUpdateUtil.reportsDirectoryPath + "/nonEWASObjectsWithCOSMICIdentifiers_"+dateSuffix+".csv"), CSVFormat.DEFAULT.withHeader("COSMIC identifier", "Referred-to-by CadidateSet (or other non-EWAS)"));
+			CSVPrinter identifiersWithNoReferrerPrinter = new CSVPrinter(new FileWriter(COSMICUpdateUtil.reportsDirectoryPath + "/COSMICIdentifiersNoReferrers_"+dateSuffix+".csv"), CSVFormat.DEFAULT.withHeader("COSMIC identifier")))
+		{
+			for (GKInstance inst : cosmicObjects)
```
Perhaps `inst` could be renamed to `cosmicObject`?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+				boolean foundMismatchedRefSequence = referenceSequenceMismatchesResidues(refSequence, modResidues);
+				
+				// If we get to the end of the loop and there is a mismatch, then COSF.
+				prefix = foundMismatchedRefSequence ? COSMIC_FUSION_PREFIX : COSMIC_LEGACY_PREFIX;
```
I'm a bit confused by this.  Won't the value of `prefix` be determined by the last loop iteration since the loop doesn't exit on finding a mismatch, so a legacy may be put even with a mismatch?

SolomonShorser-OICR replied:
I think there is generally only one EWAS, but the database always returns collections so I coded to that. I'll double-check the multi-EWAS situation, though.

SolomonShorser-OICR replied:
Hmm this may be a bug - I think it was supposed to exit the loop. Will look into this more closely.

SolomonShorser-OICR replied:
Did a test and there were no situations where a prefix "flipped" (switched from one value to another), but yes, it might be better to exit the loop early.

SolomonShorser-OICR replied:
It's a while-loop now.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+		}
+		else
+		{
+			logger.error("Wrong number of COSMIC refDBs: {} ; only 1 was expected.", refDBs.size());
+			logger.error("Cannot proceeed!");
```
Maybe logger.fatal() here?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,320 @@
+	public static void printIdentifierUpdateReport(Map<String, COSMICIdentifierUpdater> updaters) throws IOException
+	{
+		// Create the reports directory if it's missing.
+		if (!Files.exists(Paths.get(COSMICUpdateUtil.reportsDirectoryPath)))
```
Same as https://github.com/reactome/release-cosmic-update/pull/1/files#r568874126

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,254 @@
+	@Parameter(names = {"-d"}, converter = DurationConverter.class,
+			description = "Download the files, if the age of local files exceeds the specified max age, or if the files don't exist."
+					+ "Omitting this parameter means download will not occur. Specifying -d with a 0 value will force a download."
+					+ "Format for a Duration can be found here: https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#parse-java.lang.CharSequence-")
```
Maybe an example value for common duration could be added.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,254 @@
+				{
+					return !((String)inst.getAttributeValue(ReactomeJavaConstants.identifier)).toUpperCase().startsWith("COSV");
+				}
+				catch (Exception e)
+				{
+					e.printStackTrace();
+				}
```
We may want to also log the exception with a custom message for context, too?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,254 @@
+	 */
+	private void validateConfigValue(String value, String message)
+	{
+		if (value == null || value.isBlank() || value.isEmpty())
```
I think `isBlank()` only exists in Java 11+ from what I searched

jweiser commented on src/test/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdaterTest.java:
```diff
@@ -0,0 +1,128 @@
+		COSMICIdentifierUpdater updater = new COSMICIdentifierUpdater();
+		updater.setIdentifier("123456");
+		updater.setSuggestedPrefix("COSM");
+		// testing COSM so don't set a COSV identiefier.
```
Typo "identifier"

jweiser commented on src/test/java/org/reactome/release/cosmicupdate/COSMICUpdateUtilTest.java:
```diff
@@ -0,0 +1,226 @@
+		SchemaClass mockFragInsertSchemaClass = Mockito.mock(SchemaClass.class);
+		
+		// set up mock object - will not have any EWASes
+		Mockito.when(mockCosmicObject.getAttributeValue(ReactomeJavaConstants.identifier)).thenReturn("5678");
```
Could all numeric literals in this test get descriptive constant names for context and re-use?

jweiser commented on src/test/java/org/reactome/release/cosmicupdate/COSMICUpdateUtilTest.java:
```diff
@@ -0,0 +1,226 @@
+	
+	
+	@Test
+	public void testValidateAgainstFiles()
```
This test looks to be testing many different scenarios.  Could each case get a separate test with common logic in helper methods?

SolomonShorser-OICR replied:
What do you mean by "different scenarios"? The different files? I can refactor `validateIdentifiersAgainstFiles` code to call different methods - one for each file, but I think the scope of those methods should be kept private, so the only way to access them is still through `validateIdentifiersAgainstFiles`.

jweiser replied:
I just mean the different cases for the prefixes.  I think this comment was meant to apply to the `testDeterminePrefixes` test method, but I was thinking it would be good to split it into one test method for each kind of possibility - no EWASes, EWASes but with no modified residue, EWASes with modified residues, EWASes with modified residues that are fragment types etc.  

Right now everything is part of the same test and it's very long and hard to understand.

SolomonShorser-OICR committed:
Allow 1:n COSMIC identifier:object mappings

Also, break EWAS-check loop early. And some other smaller PR-requested
fixes.

SolomonShorser-OICR committed:
Minor cleanup

SolomonShorser-OICR committed:
Throw an Error.

SolomonShorser-OICR committed:
Small cleanup

SolomonShorser-OICR committed:
Should have been >= instead of <=

SolomonShorser-OICR committed:
Ensure Java 8 for build.

SolomonShorser-OICR committed:
Refactor very simple expression.

SolomonShorser-OICR committed:
Use code in org.reactome.release.jenkins.utilities.Utilities

SolomonShorser-OICR committed:
Small fix: path to prior Release step.

SolomonShorser-OICR committed:
Add SNAPSHOT repository for release-common-lib

SolomonShorser-OICR committed:
Add maven-assembly-plugin

SolomonShorser-OICR committed:
Fix path to main class.

SolomonShorser-OICR committed:
Fixed path to Main class.

SolomonShorser-OICR committed:
CLI arg for path to config.

SolomonShorser-OICR committed:
Include argument for config file.

cookersjs commented on cosmic-update.jenkinsfile:
```diff
@@ -131,15 +116,13 @@ pipeline
 			{
```
I can't actually comment above (the email stage) because technically it wasn't changed in these most recent changes, but the `${currentRelease}` variables need to be done so that `releaseVersion` is retrieved via `utils.getReleaseVersion()` and the `currentRelease` variables updated accordingly

SolomonShorser-OICR replied:
If you view the whole file, you should be able to comment on whatever line you like. And since _everything_ is new in this PR, you should be able to comment anywhere.

cookersjs replied:
Yea I think it was just because I was looking at it through the 'changes since you last reviewed' view

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -57,26 +64,49 @@ public static void main(String... args)
 		{
-			try(Reader configReader = new FileReader("src/main/resources/config.properties"))
+			boolean noConfigSpecified = Main.configPath != null && !Main.configPath.trim().isEmpty();
+//			boolean noAuthConfigSpecified = Main.authPath != null && !Main.authPath.trim().isEmpty();
```
Should this (and other lines commented out below) be removed? Unless they represent future functionality

SolomonShorser-OICR replied:
This is still WIP - this is a _draft_ review, still.

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -57,26 +64,49 @@ public static void main(String... args)
+					// These have NO default.
+					Main.COSMICUsername = configProps.getProperty("cosmic.username");
+					Main.COSMICPassword = configProps.getProperty("cosmic.password");
+					Main.personID = Long.parseLong(configProps.getProperty("personID"));
```
Could this property actually become `personId` (lowercase d), as that is how it is done through the Release project

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -57,26 +64,49 @@ public static void main(String... args)
+					Main.COSMICFusionExport = configProps.getProperty("pathToFusionExportFile", "./CosmicFusionExport.tsv");
+					Main.COSMICMutationTracking = configProps.getProperty("pathToMutationTrackingFile", "./CosmicMutationTracking.tsv");
+					// These have NO default.
+					Main.COSMICUsername = configProps.getProperty("cosmic.username");
```
In addlinks, I believe it actually uses the property name `cosmic.user` - could these property names get synced up?

SolomonShorser-OICR committed:
get release version

cookersjs commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -113,7 +143,7 @@ public void executeStep(Properties props) throws Exception
 			execService.invokeAll(Arrays.asList(unzipper1, unzipper2, unzipper3));
 			execService.shutdown();
-			
+//			
```
Remove?

SolomonShorser-OICR committed:
change userID to userId, cosmic.username to cosmic.user

SolomonShorser-OICR committed:
Clean up uncompressed files after running.

SolomonShorser-OICR committed:
Fix credentials - use file.

SolomonShorser-OICR committed:
Fix tar flags

SolomonShorser-OICR committed:
comment out other stages to only run Email stage.

SolomonShorser-OICR committed:
Fix: replace currentRelease with releaseVersion

SolomonShorser-OICR committed:
Fix archive stage.

SolomonShorser-OICR committed:
Fix archive step.

SolomonShorser-OICR committed:
Restore file.

SolomonShorser-OICR commented:
This has already run successfully in Production so I think it's _more_ than ready for review.

jweiser commented on cosmic-update.jenkinsfile:
```diff
@@ -0,0 +1,131 @@
+					def releaseVersion = utils.getReleaseVersion()
+					sh "tar -czf cosmic-update-v${releaseVersion}-reports.tgz reports/"
+					emailext (
+						body: "Hello,\n\nThis is an automated message from Jenkins regarding an update for v${releaseVersion}. The COSMIC Update step has completed. Please review the reports attached to this email. If they look correct, these reports need to be uploaded to the Reactome Drive at Reactome>Release>Release QA>V${releaseVersion}_QA>V${releaseVersion}_COSMIC_Update_Reports. The URL to the new V${release ...
```
Not a huge deal, but could this be broken up into multi lines so it's easier to read in an IDE or text editor without a lot of horizontal scrolling?

SolomonShorser-OICR replied:
This was copied from Justin. I also think it's gross, I just assumed there was a reason he did that. Maybe Jenkinsfiles don't handle multi-line string expressions? I can look into it.

jweiser replied:
Thanks!

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICIdentifierUpdater.java:
```diff
@@ -0,0 +1,194 @@
+	 * The display name of <code>identifierObject</code> will also be regenerated to reflect changes in <code>identifierValue</code>.
+	 * @param adaptor
+	 * @param modifiedForCOSMICUpdate An InstanceEdit which explains why an instance was modified.
+	 * @param identifierObject An object that represents a COSMIC identifier.
```
Could the GKInstance class type(s) the identifierObject could be, be listed for context here?

SolomonShorser-OICR replied:
There are no hard restrictions on type. It's an object that has a COSMIC identifier.

jweiser replied:
Okay. I think it's most often an EWAS, though.  Maybe an addition to the doc comment - "An object (often an EWAS, but many instance types are possible) that represents a COSMIC identifier.  Personally it helps me orient knowing what type of GKInstance I may or am likely dealing with.

SolomonShorser-OICR replied:
I think that the objects here are DatabaseIdentifier objects associated with some other object - _probably_ an EWAS.

jweiser replied:
Okay, so maybe: "An object (most likely a Database Identifier instance) that represents a COSMIC identifier"

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,362 @@
+				String genomicID = record.get(COSMIC_GENOMIC_MUTATION_ID);
+				if (updaters.containsKey(legacyID))
+				{
+					updaters.get(legacyID).forEach(updater -> {
+						updater.getMutationIDs().add(mutationID);
+						updater.setCosvIdentifier(genomicID);
+					});
```
I noticed in the other 2 validation methods, `updater.setValid(true)` is included.  Is there something different about this validation that it's not needed over here?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,362 @@
+	private static void checkEWASes(CSVPrinter nonEWASPrinter, String identifier, COSMICIdentifierUpdater updater, Collection<GKInstance> EWASes) throws InvalidAttributeException, Exception, IOException
+	{
+		String prefix;
+		GKInstance[] EWASArray = EWASes.toArray(new GKInstance[0]);
+		boolean done = false;
+		int i = 0;
+		while(!done)
+		{
```
You could get rid of the `done`, `i`, and `EWASArray` variables by using getting an iterator on the `EWASes` Collection and doing: 

`while(EWASesIterator.hasNext() && !foundMismatchedRefSequence)`

It might be easier and clearer than navigating the indexes explicitly and you can have the end condition in the while loop itself.

SolomonShorser-OICR replied:
I find the explicitness of this easier to read/comprehend.

jweiser replied:
Okay

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,362 @@
+	 * refSequence. 
+	 * @param refSequence A Reference Sequence
+	 * @param modResidues The modified residues.
+	 * @return TRUE if there is a mismatch: a mismatch is when the reference sequence DBID != the modified residues's reference sequence's DBID. FALSE, otherwise.
```
Typo - "residue's"?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -0,0 +1,362 @@
+	static Collection<GKInstance> getCOSMICIdentifiers(MySQLAdaptor adaptor) throws SQLException, Exception, InvalidAttributeException
+	{
+		@SuppressWarnings("unchecked")
+		Collection<GKInstance> refDBs = adaptor.fetchInstanceByAttribute(ReactomeJavaConstants.ReferenceDatabase, ReactomeJavaConstants.name, " = ", "COSMIC");
+		GKInstance cosmicRefDB = null;
+		if (refDBs.size() == 1)
+		{
+			cosmicRefDB = refDBs.stream().findFirst().get();
+		}
```
Maybe cast or convert the Collection to a List so streaming isn't necessary on line 295?  That way you could just do `cosmicRefDB = refDBs.get(0);`

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+					+ " Example: PT48H == \"48 hours\"")
+	private Duration fileAge;
+	
+	@Parameter(names = {"-c"}, description = "The path to the configuration file. Default is src/main/resources/config.properties")
+	private String configPath = "src/main/resources/config.properties";
```
I'm not sure if it's possible, but if `@Parameter` can use a variable name, maybe swap these two lines and say `"Default is " + configPath` in case it needs to be updated so the doc and the code values stay synced?

SolomonShorser-OICR replied:
I think an annotation applies to the next thing declared after it, so I don't think swapping the order will work. It doesn't give a compiler error, but I think that's because the compiler assumes the annotation applies to the _next_ member variable listed.

jweiser replied:
Ah okay.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+	@Parameter(names = {"-c"}, description = "The path to the configuration file. Default is src/main/resources/config.properties")
+	private String configPath = "src/main/resources/config.properties";
+	
+//	@Parameter(names = {"-a"}, description = "The path to the authorization file, containing usernames and passwords. Default is src/main/resources/auth.properties")
+//	private static String authPath = "src/main/resources/auth.properties";
```
Can or should this be removed?

SolomonShorser-OICR replied:
My plan was to have authentication values come from a separate config file. Feature is not yet complete.

jweiser replied:
Okay, sounds good.  Could a TODO be added so it's clear the commented out code is a work in progress and not something that forgot to be removed?

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+	private static final Logger logger = LogManager.getLogger();
+	private static long personId;
+
+	public static void main(String... args)
```
Just curious why using var args instead of a String array here?

SolomonShorser-OICR replied:
Why not? It's the same thing.

jweiser replied:
True, but the convention since the beginning of Java has always been to use an explicit array `main(String[] args)`.

It's not a problem, but it jumps out.  If there's no explicit reason for using var args, it might be better to stay with the convention that all Java developers recognize.

SolomonShorser-OICR replied:
I've seen `String... args` many times since var-args was first introduced. Some code generators/wizards default to this style, and have been doing so for years.

Personally, I prefer using the newer standard. I also find it more aesthetically pleasing.

jweiser replied:
Okay, I guess I just haven't seen it before.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+	private Properties getConfig()
+	{
+		boolean noConfigSpecified = this.configPath != null && this.configPath.trim().isEmpty();
+//		boolean noAuthConfigSpecified = Main.authPath != null && !Main.authPath.trim().isEmpty();
```
Can this be removed?

SolomonShorser-OICR replied:
My plan was to have authentication values come from a separate config file. Feature is not yet complete.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+	{
+		boolean noConfigSpecified = this.configPath != null && this.configPath.trim().isEmpty();
+//		boolean noAuthConfigSpecified = Main.authPath != null && !Main.authPath.trim().isEmpty();
+		if (!noConfigSpecified /* && !noAuthConfigSpecified */)
```
Would this condition need to be uncommented for a future code change?  If so, maybe a TODO block to detail the circumstances where it would be used?

SolomonShorser-OICR replied:
My plan was to have authentication values come from a separate config file. Feature is not yet complete.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+				Main.COSMICUsername = configProps.getProperty("cosmic.user");
+				Main.COSMICPassword = configProps.getProperty("cosmic.password");
+				Main.personId = Long.parseLong(configProps.getProperty("personId"));
+				// These must be set in the properties file.
+				Main.COSMICMutantExportURL = configProps.getProperty("urlToMutantExportFile");
+				Main.COSMICFusionExportURL = configProps.getProperty("urlToFusionExportFile");
+				Main.COSMICMutationTrackingURL = configProps.getProperty("urlToMutationTrackingFile");
```
Should these be checked to see if they received values and an exception thrown if not?  `getProperty` returns `null` if there isn't a value found.

SolomonShorser-OICR replied:
Re-working exception handling, in a larger scope...

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+				Main.COSMICMutantExportURL = configProps.getProperty("urlToMutantExportFile");
+				Main.COSMICFusionExportURL = configProps.getProperty("urlToFusionExportFile");
+				Main.COSMICMutationTrackingURL = configProps.getProperty("urlToMutationTrackingFile");
+//				Main cosmicUpdateStep = new Main();
+//				JCommander.newBuilder().addObject(cosmicUpdateStep).build().parse(args);
+//				cosmicUpdateStep.executeStep(configProps);
```
Can this be removed?

SolomonShorser-OICR replied:
Related to: My plan was to have authentication values come from a separate config file. Feature is not yet complete.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+//				logger.fatal("Empty/null value for path to auth.properties file was specified. This MUST be specified as \"-c /path/to/auth.properties\" "
+//						+ "or do not specify anything and the default will be src/main/resources/auth.properties");
+//			}
+			throw new Error("Empty/null values were given for the path to a config file. The program cannot run withouth the config file.");
```
I think Errors are usually reserved for something for the JVM.  Could a RuntimeException be used?

SolomonShorser-OICR replied:
No, anyone can use an Error, not just the JVM.

> An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. 

See: https://docs.oracle.com/javase/8/docs/api/java/lang/Error.html

jweiser replied:
Yes, anyone can use Error but "Effective Java" (the accepted guide book for Java best practices) recommends using RuntimeException or a subclass.  I think it is the accepted convention.

Is there a reason Error is preferred over RuntimeException here?  They have the same semantic meaning and exceptions are usually preferred in program logic unless it's a system thing like an out of memory error.

SolomonShorser-OICR replied:
Because according to the Java spec, with a RuntimeException, "recovery may still be possible". (https://docs.oracle.com/javase/specs/jls/se7/html/jls-11.html#jls-11.1.1)

I'm using the `Error` in a place where recovery _is not possible_.

---

Effective Java recommends against _subclassing_ `Error`:
> ...it's best not to implement any new Error subclasses. Therefore, all of the unchecked throwables you implement should subclass RuntimeException.
[page 244].

Effective Java also says that RuntimeExceptions should indicate programming errors. I'm not sure this particular case is a programming error: Someone ran the program and did not specify the path to the configuration file. That seems like a user error, in the sense that the user ran the program with incorrect inputs.

jweiser replied:
I think the context of "recovery is not possible" refers to an environmental problem like a memory issue as opposed to a problem with the program state (I think this is what is meant by "programming errors" - an issue to with the program and/or its input/state) where handling it is possible even if it's not desirable, as in this case.  I still think an exception may be better as it gives the caller the possibility of doing something with the issue where are error doesn't and it can be thrown up the stack to terminate the program or caught explicitly and the program terminated explicitly.

It's a bit of hair splitting in terms of Error vs Exception, so I'm okay if you want to keep this as is, but I do think Exception is better for handling program/state/logic related issues.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+				{
+					// exception caught here means there is probably some fundamental problem with the data
+					// such that the program should probably not continue, so throw Error.
+					throw new Error(e);
```
Same as https://github.com/reactome/release-cosmic-update/pull/1/files#r578640453

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+		}
+	}
+
+	/** Cean up the uncompressed data files. Don't remove the zipped files, because if you need to run this code again,
```
Typo - "Clean"

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+		}
+		catch (URISyntaxException e)
+		{
+			e.printStackTrace();
+			throw e;
```
Is a stacktrace needed here if the exception is rethrown and dealt with later?

SolomonShorser-OICR replied:
Exception handling is being re-worked...

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+	 */
+	private static void updateIdentifiers(MySQLAdaptor adaptor, Map<String, List<COSMICIdentifierUpdater>> updates) throws Exception
+	{
+		for (List<COSMICIdentifierUpdater> updater : updates.values())
```
Could `updater` be renamed `updaterList` since it is more than one updater?

SolomonShorser-OICR replied:
Yes, I think the name remained after several rounds of refactoring.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -0,0 +1,337 @@
+	{
+		for (List<COSMICIdentifierUpdater> updater : updates.values())
+		{
+			updater.forEach(u -> {
```
If `updater` is renamed `updaterList`, it would help to rename `u` as `updater` since here it is iterating over each one individually.

jweiser commented on src/test/java/org/reactome/release/cosmicupdate/COSMICUpdateUtilTest.java:
```diff
@@ -0,0 +1,236 @@
+	private static final String IDENTIFIER_1234 = "1234";
+	private static final long DBID_1235 = 1235L;
+	private static final long DBID_1234 = 1234L;
+	private static final String IDENTIFIER_COSM1111 = "COSM1111";
+	private static final String IDENTIFIER_44444 = "44444";
+	private static final String IDENTIFIER_5678 = "5678";
+	private static final String IDENTIFIER_COSF6321 = "COSF6321";
+	private static final String IDENTIFIER_COSF7890 = "COSF7890";
```
Thanks for adding constants for these!  Instead of the dummy value in the name, could the identifier's purpose be stated.  What is the difference between the usage of the identifiers "44444" and "5678", for example?

SolomonShorser-OICR replied:
They're just random test numbers, no deeper purpose. By putting the value into the name, I know exactly what it is without having to refer back to the variable declaration.

jweiser replied:
Okay.  I won't push it, but if they don't have meaning you could consider naming them something like "DUMMY_IDENTIFIER_1", "DUMMY_IDENTIFIER_2", "DUMMY_DB_ID_1", "DUMMY_COSM_1", etc. so it's clear the values assigned just need to be distinct and don't have a meaning.

I'll leave that to your discretion, though.

jweiser commented on src/test/java/org/reactome/release/cosmicupdate/COSMICUpdateUtilTest.java:
```diff
@@ -0,0 +1,236 @@
+		Path pathToMutationTracking = Files.createTempFile("mutationTracking", "csv");
+		this.COSMICMutationTrackingFile = pathToMutationTracking.toString();
+		Files.writeString(pathToMutationTracking, MUTANT_HEADER);
+		Files.writeString(pathToMutationTracking, "1234\t1234\tCOSV2323232\n", StandardOpenOption.APPEND);
```
Could the "IDENTIFIER_1234" constant be used here instead of hard-coding the "1234" value?

jweiser commented on src/test/java/org/reactome/release/cosmicupdate/COSMICUpdateUtilTest.java:
```diff
@@ -0,0 +1,236 @@
+		Path pathToMutantExport = Files.createTempFile("mutantExport", "csv");
+		this.COSMICMutantExportFile = pathToMutantExport.toString();
+		Files.writeString(pathToMutantExport, MUTANT_HEADER);
+		Files.writeString(pathToMutantExport, "1234\t1234\tCOSV2323232\n", StandardOpenOption.APPEND);
```
Same as https://github.com/reactome/release-cosmic-update/pull/1/files#r578651248

jweiser reviewed (changes requested):
Overall looks good.  The only thing I saw that I'd like confirmation on is this:

https://github.com/reactome/release-cosmic-update/pull/1/files#r578619747

I just want to make sure that's not an omission.

Everything else is style suggestions and not blockers.

SolomonShorser-OICR committed:
Fix process to update displayName

SolomonShorser-OICR committed:
Improve documentation, mostly.

SolomonShorser-OICR committed:
Better exception handling.

SolomonShorser-OICR commented:
I've added some comments explaining a bit more detail about how the validation works.

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/COSMICUpdateUtil.java:
```diff
@@ -66,6 +66,20 @@ private COSMICUpdateUtil()
+		//         and indicate that the identifier is VALID. 
+		// 
+		// The goal is to set valid = true for an identifier that is currently valid in COSMIC - if you search that identifier, or create a link containing it, you will get 
+		// a result. Sometimes, there is no mapping to a COSV identifier for a given COSM identifier. This will result in "valid == fale".
```
Typo - "false"

jweiser commented on src/main/java/org/reactome/release/cosmicupdate/Main.java:
```diff
@@ -142,61 +134,101 @@ private Properties getConfig()
 			}
-			cleanupFiles();
+		}
+		// These exceptions are all unrecoverable, but they get a differen catch block so we can give the user a better message.
```
Typo - "different"

SolomonShorser-OICR committed:
Minor updates.

SolomonShorser-OICR committed:
Minor fixes.
